### PR TITLE
"Using application configuration for libraries" added (Anti-patterns documentation)

### PR DESCRIPTION
**Pull request by lucasvegi**

### CODE REVIEW

**Review by whatyouhide — COMMENTED** [review]
One suggestion I have is to simulate a slightly *more complex* example that however can be significantly more realistic.

My go to is once again something where the antipattern is configuring for example a base URL through app config

```elixir
config :api_client,
  base_url: "..."
```

and the refactoring would be to pass it as an option:

```elixir
APIClient.request(..., base_url: "...")
```

File: lib/elixir/pages/anti-patterns/design-anti-patterns.md [hunk 1]
```diff
@@ -279,4 +279,59 @@ A common practice followed by the community is to make the non-raising version t
...
+```elixir
+iex> DashSplitter.split("Lucas-Francisco-Vegi")
+["Lucas", "Francisco", "Vegi"]
+iex> DashSplitter.split("Lucas-Francisco-da-Matta-Vegi")
+["Lucas", "Francisco", "da-Matta-Vegi"]
+```
+
+#### Refactoring
+
+To remove this anti-pattern and make the library more adaptable and flexible, this type of configuration must be performed via parameters in function calls. The code shown below performs the refactoring of the `split/1` function by adding a new optional parameter of type *Keyword list*. With this new parameter, it is possible to modify the default behavior of the function at the time of its call, allowing multiple different ways of using `split/2` within the same application:
```
[the remark this answers is not in the document]
Comment (wojtekmach) [reply lib/elixir/pages/anti-patterns/design-anti-patterns.md:321]: `` [keyword list](`t:keyword/0`) `` would work. it wouldn't be a very useful link however, it goes to the huge list of types at https://hexdocs.pm/elixir/typespecs.html#built-in-types.

[the remark this answers is not in the document]
Comment (eksperimental) [reply lib/elixir/pages/anti-patterns/design-anti-patterns.md:321]: True, then we could link to `t:Keyword.t/0`.

**Comment by lucasvegi** [comment]
@whatyouhide What do you think of this example? If it's cool I'll adapt the text...

```elixir
import Config

config :api_client,
  base_url: "https://api.github.com"

import_config "#{config_env()}.exs"
```

```elixir
defmodule APIClient do
  def request(endpoint) when is_binary(endpoint) do
    HTTPoison.start
    base_url = Application.fetch_env!(:api_client, :base_url)  # <= retrieve parameterized value
    HTTPoison.get(base_url <> endpoint)                        # <= base_url: "https://api.github.com"
  end
end
```

```elixir
#...Use examples with anti-pattern

iex> APIClient.request("/users/lucasvegi")  #<== OK!
{:ok,
 %HTTPoison.Response{
   status_code: 200,
   body: "...",
   headers: [ ... ],
   request_url: "https://api.github.com/users/lucasvegi",
   ...
 }}

iex> APIClient.request("/persons?_quantity=1")  #<== Failed attempt to access a different API!
{:ok,
 %HTTPoison.Response{
   status_code: 404,
   body: "{\"message\":\"Not Found\",\"documentation_url\":\"https://docs.github.com/rest\"}",
   headers: [ ... ],
   request_url: "https://api.github.com/persons?_quantity=1",  # <== base_url never changes!
   ...
 }}
```

Refactoring:

```elixir
defmodule APIClient do
  def request(endpoint, options \\ []) when is_binary(endpoint) and is_list(options) do
    HTTPoison.start
    base_url = Keyword.get(options, :base_url, "https://api.github.com")  # <= default config of base_url == "https://api.github.com"
    HTTPoison.get(base_url <> endpoint)
  end
end
```

```elixir
#...Use examples of the refactored code

iex> APIClient.request("/users/lucasvegi")  #<== OK with default config!
{:ok,
 %HTTPoison.Response{
   status_code: 200,
   body: "...",
   headers: [ ... ],
   request_url: "https://api.github.com/users/lucasvegi",
   ...
 }}

iex> APIClient.request("/persons?_quantity=1", [base_url: "https://fakerapi.it/api/v1"]) #<== OK with another API!
{:ok,
%HTTPoison.Response{
  status_code: 200,
  body: "{...\"data\":[{\"id\":1,\"firstname\":\"Taryn\",\"lastname\":\"Price\",...}",
  headers: [ ... ],
  request_url: "https://fakerapi.it/api/v1/persons?_quantity=1",
  ...
}}

**Comment by whatyouhide** [comment]
@lucasvegi yes that's exactly what I had in mind. Few notes:

  * We don't need to mention HTTPoison specifically. Let's replace it with `MyApp.HTTPClient`.
  * We don't need to be verbose when showing return values as you do in your example, as it's not what we're talking about here.

Does that make sense?

**Comment by whatyouhide** [comment]
@lucasvegi yep!

**Comment by lucasvegi** [comment]
> @lucasvegi yep!

Great! I'll make some adjustments and commit

File: lib/elixir/pages/anti-patterns/design-anti-patterns.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/elixir/pages/anti-patterns/design-anti-patterns.md:321]:
```suggestion
To remove this anti-pattern and make the library more adaptable and flexible, this type of configuration must be performed via parameters in function calls. The code shown below performs the refactoring of the `split/1` function by accepting [keyword lists](`Keyword`) as a new optional parameter. With this new parameter, it is possible to modify the default behavior of the function at the time of its call, allowing multiple different ways of using `split/2` within the same application:
```

File: lib/elixir/pages/anti-patterns/design-anti-patterns.md [hunk 2]
```diff
@@ -279,4 +279,59 @@ A common practice followed by the community is to make the non-raising version t
 
 ## Using application configuration for libraries
 
-TODO
+#### Problem
+
+The [*application environment*](https://hexdocs.pm/elixir/1.15/Application.html#module-the-application-environment) can be used to parameterize global values that can be used in an Elixir system. This parameterization mechanism can be very useful and therefore is not considered an anti-pattern by itself. However, library authors should avoid using the application environment to configure their library. The reason is exactly that the application environment is a **global** state, so there can only be a single value for each key in the environment for an application. This makes it impossible for multiple applications depending on the same library to configure the same aspect of the library in different ways.
```
Comment (josevalim) [inline lib/elixir/pages/anti-patterns/design-anti-patterns.md:284]:
```suggestion
The [*application environment*](https://hexdocs.pm/elixir/Application.html#module-the-application-environment) can be used to parameterize global values that can be used in an Elixir system. This mechanism can be very useful and therefore is not considered an anti-pattern by itself. However, library authors should avoid using the application environment to configure their library. The reason is exactly that the application environment is a **global** state, so there can only be a single value for each key in the environment for an application. This makes it impossible for multiple applications depending on the same library to configure the same aspect of the library in different ways.
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**